ls1intum/Artemis
# `Development`: Refactor quiz exercise service

### Checklist
#### General


#### Server

### Motivation and Context



The implementation of Quiz Pool API will reuse the functionalities of Quiz Exercise API. Therefore, QuizExerciseService is refactored so that the functionality can be reused later by Quiz Pool.

### Description


1. Create QuizService class
QuizService is an abstract class that contains the common methods of QuizExerciseService and QuizPoolService (will be implemented on separate PR).
2. Move the save, saveCorrectMappingsInIndices, saveCorrectMappingsInIndicesShortAnswer, restoreCorrectMappingsFromIndices, and restoreCorrectMappingsFromIndicesShortAnswer methods to QuizService.
The methods, which are used by QuizExerciseService, will be used by the QuizPoolService as well.
3. Create QuizConfiguration Interface
QuizConfiguration is an abstraction that contains the common operation of QuizExercise and QuizPool (will be implemented on separate PR)
4. Move the reconnectJSONIgnoreAttributes method to QuizConfiguration
The reconnectJSONIgnoreAttributes method, which is used by QuizExercise, will be used by the QuizPool as well.
5. Refactor QuizConfiguration.reconnectJSONIgnoreAttributes and QuizService.save method

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Course

1. Log in to Artemis
2. Navigate to Course Administration
3. Select the Course
7. Navigate to Exercises
8. Create a new Quiz Exercise
9. Verify that the Quiz Exercise can be created successfully.
10. Edit the newly created Quiz Exercise
11. Verify that the Quiz Exercise can be updated successfully.
12. Verify that the Quiz Exercise can be reevaluated successfully.

### Review Progress






#### Code Review
#### Manual Tests

### Test Coverage



| Class/File | Branch | Line |
|------------|-------:|-----:|
| QuizConfiguration.java | 78% | 100% |
| QuizExercise.java | 73% | 91% |
| QuizExerciseService.java | 92% | 100% |
| QuizService.java | 82% | 94% |
| AnswerCounter.java | n/a | 100% |
| AnswerOption.java | 50% | 94% |
| DragAndDropMapping.java | 50% | 90% |
| DragAndDropQuestion.java | 49% | 83% |
| DragItem.java | 33% | 82% |
| DropLocation.java | 58% | 94% |
| DropLocationCounter.java | n/a | 100% |
| MultipleChoiceQuestion.java | 62% | 88% |
| QuizQuestion.java | 66% | 96% |
| QuizQuestionStatistic.java | n/a | 92% |
| ShortAnswerMapping.java | 75% | 97% |
| ShortAnswerQuestion.java | 55% | 87% |
| ShortAnswerSolution.java | 50% | 93% |
| ShortAnswerSpot.java | 75% | 97% |
| ShortAnswerSpotCounter.java | n/a | 100% |

### Screenshots



https://user-images.githubusercontent.com/10885503/230736567-c00948a3-d60f-4a82-af8a-fa79f23c7643.mov
